File: 프로그래머스/2/17679. ［1차］ 프렌즈4블록/［1차］ 프렌즈4블록.py
def solution(m, n, board):
    cnt = 0
    # data 정의
    data = [[0]*n for _ in range(m)]
    for i in range(m):
        for j in range(n):
            data[i][j] = [board[i][j], False]
    while True:
        # 2x2 블록 제거
        temp_cnt = 0
        for i in range(m-1):
            for j in range(n-1):
                if data[i][j][0] == data[i+1][j][0] == data[i][j+1][0] == data[i+1][j+1][0] != 'X':
                    data[i][j][1] = True
                    data[i+1][j][1] = True
                    data[i][j+1][1] = True
                    data[i+1][j+1][1] = True
        for i in range(m):
            for j in range(n):
                if data[i][j][0] != 'X' and data[i][j][1] == True:
                    data[i][j] = ['X', False]
                    temp_cnt += 1
        cnt += temp_cnt

        if temp_cnt == 0:
            break
        # 없어진 블록 채우기
        for _ in range(m-1):  # 최대 m-1번
            for j in range(n):
                for i in range(m-2, -1, -1):
                    if data[i+1][j] == ['X', False]:
                        data[i+1][j], data[i][j] = data[i][j], data[i+1][j]

    return cnt

File: 프로그래머스/2/17679. ［1차］ 프렌즈4블록/［1차］ 프렌즈4블록.java
import java.util.Arrays;

class Solution {
    public int solution(int m, int n, String[] board) {
        char[][] map = new char[m][n];
        for (int i = 0; i < m; i++) {
            map[i] = board[i].toCharArray();
        }

        while (true) {
            // 임시 맵 복사
            char[][] tmp = new char[m][n];
            for (int i = 0; i < m; i++) {
                tmp[i] = map[i].clone();
            }

            // 2x2 모노톤 찾기
            boolean found = false; // 블록이 발견되었는지 여부
            for (int i = 0; i < m - 1; i++) {
                for (int j = 0; j < n - 1; j++) {
                    char target = map[i][j];

                    if (target == ' ') {
                        continue;
                    }

                    char right = map[i][j + 1];
                    char down = map[i + 1][j];
                    char rightDown = map[i + 1][j + 1];

                    if (target == right && target == down && target == rightDown) {
                        tmp[i][j] = ' ';
                        tmp[i][j + 1] = ' ';
                        tmp[i + 1][j] = ' ';
                        tmp[i + 1][j + 1] = ' ';
                        found = true; // 블록이 발견됨
                    }
                }
            }

            // 아무 블록도 발견되지 않으면 종료
            if (!found) {
                break;
            }

            // 밑으로 내리기
            for (int col = 0; col < n; col++) {
                for (int row = m - 1; row >= 0; row--) {
                    if (tmp[row][col] == ' ') {
                        for (int k = row - 1; k >= 0; k--) {
                            if (tmp[k][col] != ' ') {
                                tmp[row][col] = tmp[k][col];
                                tmp[k][col] = ' ';
                                break;
                            }
                        }
                    }
                }
            }

            // 맵 업데이트
            map = tmp;
        }

        return count(map);
    }

    private int count(char[][] map) {
        int cnt = 0;
        for (int i = 0; i < map.length; i++) {
            for (int j = 0; j < map[0].length; j++) {
                if (map[i][j] == ' ') {
                    cnt++;
                }
            }
        }
        return cnt;
    }
}
